src/ddspdrum/module.py

```diff
@@ -0,0 +1,239 @@
+"""
+Synth modules.
+
+    TODO    :   - VCA is slow. Blocks? Not sure what's best for tensors.
+                - Convert operations to tensors, obvs.
+"""
+
+import numpy as np
+from scipy.signal import resample
+
+from ddspdrum.defaults import CONTROL_RATE, SAMPLE_RATE
+
+
+class SynthModule:
+    """
+    Base class for synthesis modules. Mostly helper functions for the moment.
+    """
+
+    def __init__(
+        self, sample_rate: int = SAMPLE_RATE, control_rate: int = CONTROL_RATE
+    ):
+        self.sample_rate = SAMPLE_RATE
+        self.control_rate = CONTROL_RATE
+
+    def control_to_sample_rate(self, control: np.array) -> np.array:
+        """
+        Resample a control signal to the sample rate
+        TODO: One thing I worry about with all these casts to
+        ints back and forth between sample and control rate is
+        off-by-one errors.
+        I'm beginning to believe we should standardize as much
+        as possible on nsamples in most places instead of
+        float duration in seconds, and just convert to ncontrol
+        when needed..
+        """
+        # Right now it appears that all signals are 1d, but later
+        # we'll probably convert things to 2d: instance x signal
+        assert control.ndim == 1
+        nsamples = int(round(len(control) * self.sample_rate / self.control_rate))
+        return resample(control, nsamples)
+
+    def fix_length(self, signal: np.array, length: int) -> np.array:
+        # Right now it appears that all signals are 1d, but later
+        # we'll probably convert things to 2d: instance x signal
+        assert signal.ndim == 1
+        if len(signal) < length:
+            signal = np.pad(signal, [0, length - len(signal)])
+        elif len(signal) > length:
+            signal = signal[:length]
+        assert signal.shape == (length,)
+        return signal
+
+
+class ADSR(SynthModule):
+    """
+    Envelope class for building a control rate ADSR signal
+    """
+
+    def __init__(
+        self,
+        a: float = 0.25,
+        d: float = 0.25,
+        s: float = 0.5,
+        r: float = 0.5,
+        alpha: float = 3.0,
+        sample_rate: int = SAMPLE_RATE,
+        control_rate: int = CONTROL_RATE,
+    ):
```

Comment:
Okay, so one of the major changes is that we have to do away with this `get` and `set` nonsense. Having weird-side-effects makes thing harder to understand and also harder to port to neural networks.

We should be trying to write in a more *functional* way, without side-effects. It makes the code cleaner, and also neural networks are basically tensor functions.

Reply:
Sure thing.